CSCfi/rems
# Application expiration improvements

implements #3225 

Fixes
- Expiration notification command uses `now` for calculating `:expires-on`
  - Previously it used last applying user event for calculations, which could set the date in emails in the past.
- Send expiration notifications no longer validates invalid expiration
  - `:last-activity` was no longer returned by command, so validation always failed
  - In addition, validation returned faulty key `:error` which was ignored by (command!), see detailed description below

Feature additions/changes
- NB: (time-before-or-equal?) is used in both (expire-application) and (should-send-notification-email?)
  - Combines (time/before?) and (time/equal?) using logical-or
- NB: both `:application.command/delete` and `:application.command/send-expiration-notifications` command handlers validate that application state is draft
- (rems.application.eraser/process-applications!) now logs more:
  - **info** on startup,
  - **warning** when expirer-bot does not exist,
  - **info** when it has nothing to execute,
  - **info** before running each command, and
  - **warning** when command validation fails.
- (rems.db.applications/delete-application!) logs info after it is finished deleting application.
- Delete command now includes optional `:expires-on` similar to send expiration notifications.
  - It is for logging purposes, not used in event generation, and as optional key it can be passed from API, but has no effect.
- New config option `:application-expiration-process-limit`:
  - Limits how many commands (rems.application.eraser/process-applications!) can pull for execution.
  - Default value `nil` (hard-coded to 50)

Test additions/changes
- rems.application.expirer-bot
  - unit tests have more (testing) blocks
  - use (exactly-90d-ago) format over (over-90d-ago) due to change in date time checking, which should be more intuitive
- rems.application.test-eraser
  - all tests generally run (process-applications!) and check following:
    1. any applications removed?
    2. any notification events created?
    3. any emails sent? 
  - use clojure.tools.logging.test instead of atom + with-redefs over log function. with-log automatically handles log collection start/end
  - (testing) blocks handle as much config as possible, e.g. with-redefs, with-log and with-fixed-time
    - Due to progressive time in notification tests, fixed time needs to be redefined. also with-log needs to used for each (process-applications!), so there is already some unavoidable nesting

Send expiration notifications command had interesting bug:
1. A validation function generated error due to non-existing attribute, but it was malformed due to typo (`:error` should've been `:errors`).
4. `rems.application.commands/handle-command` ignores this because command results are not validated.
5. Now `:events` is not in cmd result, but because `:errors` is neither, `rems.service.command/command!` assumes it is valid.
6. `command!` calls `mapv` on nil, which returns empty vector, so no events are added or validated.

# Checklist for author

Remove items that aren't applicable, check items that are done.

## Reviewability
- [x] Link to issue

## Documentation
- [x] Update changelog if necessary

## Testing
- [x] Complex logic is unit tested
- [x] Valuable features are integration tested automatically

## Follow-up
- [ ] New tasks are created for pending or remaining tasks

---
## Example runtime log
Some Logback MDC parts omitted for brevity:
```log
INFO  rems.application.eraser - :start #'rems.application.eraser/process-applications!
INFO  rems.application.eraser - Total of 4 applications due for processing (4 expired, 0 to send expiration notifications for). Processing is limited to 3 applications at a time.
INFO  rems.application.eraser - :application.command/delete {:application-id 16, :expires-on #clj-time/date-time "2023-08-15T19:10:28.985Z"}
INFO  rems.db.applications - Finished deleting application 16
INFO  rems.application.eraser - :application.command/delete {:application-id 10, :expires-on #clj-time/date-time "2023-08-21T19:10:28.424Z"}
INFO  rems.db.applications - Finished deleting application 10
INFO  rems.application.eraser - :application.command/delete {:application-id 18, :expires-on #clj-time/date-time "2023-08-13T19:10:29.166Z"}
INFO  rems.db.applications - Finished deleting application 18
INFO  rems.db.applications - Start rems.db.applications/reload-cache!
INFO  rems.db.applications - Finished rems.db.applications/reload-cache!
INFO  rems.application.eraser - :finish #'rems.application.eraser/process-applications!
```

Command validation failure (e.g. configured expire for non-draft):
```log
INFO  rems.application.eraser - :start #'rems.application.eraser/process-applications!
INFO  rems.application.eraser - Total of 1 applications due for processing (1 expired, 0 to send expiration notifications for). Processing is limited to 3 applications at a time.
INFO  rems.application.eraser - :application.command/delete {:application-id 24, :expires-on #clj-time/date-time "2023-11-10T19:10:29.809Z"}
WARN  rems.application.eraser - Command validation failed: :application.command/delete {:application-id 24} {:errors [{:type :only-draft-may-be-deleted}]}
INFO  rems.db.applications - Start rems.db.applications/reload-cache!
INFO  rems.db.applications - Finished rems.db.applications/reload-cache!
INFO  rems.application.eraser - :finish #'rems.application.eraser/process-applications!
```

Nothing to process, no cache reload:
```log
INFO  rems.application.eraser - :start #'rems.application.eraser/process-applications!
INFO  rems.application.eraser - No applications to process
INFO  rems.application.eraser - :finish #'rems.application.eraser/process-applications!
```
